java-checks/src/main/java/org/sonar/java/checks/FinalClassCheck.java

```diff
@@ -41,12 +42,23 @@
 
   @Override
   public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
     ClassTree classTree = (ClassTree) tree;
-    if (hasOnlyPrivateConstructors(classTree) && !ModifiersUtils.hasModifier(classTree.modifiers(), Modifier.FINAL)) {
+    if (hasOnlyPrivateConstructors(classTree) && !isExtended(classTree) && !ModifiersUtils.hasModifier(classTree.modifiers(), Modifier.FINAL)) {
       reportIssue(classTree.simpleName(), "Make this class \"final\" or add a public constructor.");
     }
   }
 
+  private static boolean isExtended(ClassTree classTree) {
+    List<IdentifierTree> usages = classTree.symbol().usages();
+    return usages.stream().anyMatch(usage -> {
+      Tree parent = usage.parent();
+      return parent != null && parent.is(Kind.CLASS);
```

Comment:
This won't work if fully qualified name is used in the extend clause.
```
class Foo extends my.pack.age.AbstractClass {} 
```
I would more go for a subvisitor looking for type of extend clause and anonymous constructor.